Refactor HeadLogic with types for each state

ch1bo opened the pull request:
<!-- Describe your change here -->
:broom: Address some XXX in `HeadLogic` by creating actual types for `OpenState` et al. This make the interface between `update` and `on...` functions smaller and could be used to split the `update` into multiple cases (one for each state) later.

ch1bo committed:
Sub-type idle state

This allows us to define type-safe functions which only act on one of
the head states. Idle in this case.

ch1bo committed:
Create InitialState type in HeadLogic

ch1bo committed:
Refactor onInitialClientCommit

ch1bo committed:
Refactor onInitialChainCommitTx

ch1bo committed:
Refactor onInitialClientAbort

ch1bo committed:
Treat commit txs in non-Initial state as errors (fall-through)

ch1bo committed:
Refactor onInitialChainCollectTx

ch1bo committed:
Create OpenState type in HeadLogic

ch1bo committed:
Refactor onOpenClientClose

ch1bo committed:
Refactor onOpenClientNewTx

ch1bo committed:
Refactor onOpenNetworkReqTx

ch1bo committed:
Refactor onOpenNetworkReqSn

ch1bo committed:
Refactor onOpenNetworkAckSn

ch1bo committed:
Refactor onOpenChainCloseTx

This also fixes an (in-flight of this branch) broken field access.

Also, the TODO seems not relevant anymore so I removed it.

ch1bo committed:
Create ClosedState type in HeadLogic

This allowed us already to remove incomplete record updates from
HeadLogic module.

ch1bo committed:
Refactor onClosedChainContestTx

ch1bo committed:
Refactor onClosedClientFanout

ch1bo committed:
Refactor onClosedChainFanoutTx

ch1bo committed:
Limit arbitrary values of HeadStates to possible

We could think of encoding the possible previous states for each state
type, but this would be a bit too much right now.

Instead, this commit is motivated to limit the length of state linked
lists we generate.

v0d1ch commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1,7 +1,6 @@
 {-# LANGUAGE DuplicateRecordFields #-}
 {-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE UndecidableInstances #-}
-{-# OPTIONS_GHC -Wno-incomplete-record-updates #-}
```
Nice!

v0d1ch commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -730,85 +800,76 @@ onOpenNetworkAckSn
+  NewState closedState $
+    notifyClient :
+      [ OnChainEffect
+        { -- REVIEW: Was using "old" chainState before
```
Is using the `newChainState` here correct as we were using the previous one before?

ch1bo replied:
Not sure yet, that's why I left this comment 🙂 It might have been an oversight and is more evident now

ch1bo committed:
Update logs.yaml json schema to reflect new HeadLogic state structure

We chose not to manually implement ToJSON/FromJSON instances for
HeadState just to keep backward compatibility on the logs JSON schema.

A manual instance would mean higher maintainability demands, whereas the
logs schema is not so important as for example the api schema.

ch1bo commented on hydra-node/json-schemas/logs.yaml:
```diff
@@ -589,98 +589,135 @@ definitions:
         required:
           - tag
-          - chainState
+          - contents
```
See https://github.com/input-output-hk/hydra/pull/725/commits/54004cc2b3b82e7edc115b3b6884cda4bc903b63. Not sure if it's worth it though as these are "only in the logs" schema :/

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -689,7 +690,7 @@ withHydraNode ::
   outputs <- atomically newTQueue
   outputHistory <- newTVarIO mempty
-  nodeState <- createNodeState $ IdleState{chainState = SimpleChainState{slot = ChainSlot 0}}
+  nodeState <- createNodeState $ Idle IdleState{chainState = SimpleChainState{slot = ChainSlot 0}}
```
I'd rather not. The smart constructors would need to use positional arguments or yet another type that is isomorphic to `IdleState` et al.

ch1bo replied:
My commment applies under the assumption you mean a function like `mkIdleState :: ChainStateType tx -> HeadState tx` as "smart constructor". So this would read `nodeState <- createNodeState $ mkIdleState SimpleChainState{slot = ChainSlot 0}`.

Maybe I misunderstood you and I would need more concrete advice to consider your suggestion :)
